tensorflow_recommenders_addons/dynamic_embedding/python/ops/dynamic_embedding_optimizer.py

```diff
@@ -430,74 +524,334 @@ def apply_gradients(grads_and_vars, global_step=None, name=None):
+      for g_and_v in sparse_grads_and_vars_unaggregated:
+        grads_and_vars.append(g_and_v)
+
+    return super(keras_OptimizerV2, self).apply_gradients(grads_and_vars,
+                                                          name=name)
+
+  def compute_gradients_horovod_wrapper(compute_gradients_horovod_func):
+
+    def compute_gradients_horovod_wrapper_impl(*args, **kwargs):
+      if self._finish_register_tw is False:
+        var_list = kwargs.get('var_list', None)
+        if var_list is None:
+          raise Exception(
+              "var_list parameter must not be None when using compute_gradients function!"
+          )
+        _, trainable_wrapper_vars = trainable_wrapper_filter(var_list)
+        for tw in trainable_wrapper_vars:
+          self.register_local_var(tw)
+        self._finish_register_tw = True
+      return compute_gradients_horovod_func(*args, **kwargs)
+
+    return compute_gradients_horovod_wrapper_impl
+
   if isinstance(self, optimizer.Optimizer):
+    # TF1 Optimizer
     self._get_or_make_slot = _get_or_make_slot
     self._get_or_make_slot_with_initializer = _get_or_make_slot_with_initializer
     self._zeros_slot = _zeros_slot
-    if self._hvd_sync:
-      self.apply_gradients = apply_gradients
-  elif isinstance(self, optimizer_v2.OptimizerV2) or isinstance(
-      self, keras_optimizer):
-    if self._hvd_sync:
+    # TODO(MoFHeka): Support distribute strategy with tf.compat.v1.train.Optimizer
+    # self.get_slot = get_slot_v1
+    if self._custom_sync:
+      self.apply_gradients = apply_gradients_sync_v1
+    elif 'horovod' in str(self):
+      # If use Horovod keras optimizer wrapper
+      if not hasattr(self, 'register_local_var'):
+        import horovod
+        raise Exception(
+            f"Input optimizer must has register_local_var function! Your Horovod version is {horovod.__version__}, please upgrade to greater than 0.26.0"
+        )
+      self._finish_register_tw = False
+      compute_gradients_horovod_v1 = self.compute_gradients
+      self.compute_gradients = compute_gradients_horovod_wrapper(
+          compute_gradients_horovod_v1)
+    if distribute_ctx.has_strategy():
       raise Exception(
-          "OptimizerV2 didn't support distributed sync train now, please use tf.train.XxxxOptimizer."
+          "Distribute strategy isn't supported when using tf.compat.v1.train.Optimizer in TFRA for now."
       )
+  elif (isinstance(self, optimizer_v2_legacy.OptimizerV2)
+        or isinstance(self, keras_OptimizerV2_legacy)
+        or isinstance(self, keras_OptimizerV2)):
+    # TF2 Keras Optimizer
     if hasattr(self, '_distributed_apply'):
+      # Compatible with Keras < 2.12.0 legacy optimizer
       self.add_slot = add_slot
+      self.get_slot = get_slot_v2_lagacy
       self._distributed_apply = _distributed_apply
-    else:
+      if self._custom_sync:
+        self.apply_gradients = apply_gradients_sync_v2_lagacy
+      elif 'horovod._keras' in str(self):
```

Comment:
Is this judgment condition stable and reliable? Do we have a better method to judge the type of the optimizer?

Reply:
This method is not stable, but it could be reliable over a certain period of time. And this is judgment of Horovod:
```python
if version.parse(keras.__version__) < version.parse("2.11"):
    optimizer_class = keras.optimizers.Optimizer
else:
    optimizer_class = keras.optimizers.legacy.Optimizer
```
I don't know which one is better.

Reply:
Emm... I prefer Horovod's method.

Reply:
Yes...Horovod's method looks elegance. But sometimes there would be a different version style, such as "tensorflow==2.9+nv", "keras==2.11+dev", and so on. They may cause misjudgment.

Reply:
Make sense to me.